rllib/algorithms/algorithm.py

```diff
@@ -459,6 +466,11 @@ def __init__(
         # Placeholder for our LearnerGroup responsible for updating the RLModule(s).
         self.learner_group: Optional["LearnerGroup"] = None
 
+        # The Algorithm's `MetricsLogger` object to collect stats from all its
+        # components (including timers, counters and other stats in its own
+        # `training_step()` and other methods) as well as custom callbacks.
+        self.metrics = MetricsLogger()
```

Comment:
Correct, Algorithm as well as all of its sub-components have their own MetricsLogger object.

Thus, if Algorithm calls n of its sub-components (in parallel) and receives n result dicts from these components, each of which have already been reduced by these components, it can now put these n result dicts into its own MetricsLogger (and then reduce these further until a single result dict comes out from this). So it's a funnel like design with two funnel-axes: 1) time (a component logging n different values in a sequence under the same key) and 2) parallel/scaling (a component having n sub-components and requesting them in parallel to deliver one result dict each, then merging/reducing these n result dicts into one).